src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ImplementIDisposableCorrectly.cs

```diff
@@ -106,13 +112,494 @@ public abstract class ImplementIDisposableCorrectlyAnalyzer : DiagnosticAnalyzer
+            }
+
+            /// <summary>
+            /// Checks rule: Remove the finalizer from type {0}, override Dispose(bool disposing), and put the finalization logic in the code path where 'disposing' is false.
+            /// </summary>
+            private static void CheckFinalizeOverrideRule(INamedTypeSymbol type, SymbolAnalysisContext context)
+            {
+                if (type.HasFinalizer())
+                {
+                    context.ReportDiagnostic(type.CreateDiagnostic(FinalizeOverrideRule, type.Name));
+                }
+            }
+
+            /// <summary>
+            /// Checks rule: Provide an overridable implementation of Dispose(bool) on {0} or mark the type as sealed. A call to Dispose(false) should only clean up native resources. A call to Dispose(true) should clean up both managed and native resources.
+            /// </summary>
+            private static void CheckProvideDisposeBoolRule(INamedTypeSymbol type, SymbolAnalysisContext context)
+            {
+                context.ReportDiagnostic(type.CreateDiagnostic(ProvideDisposeBoolRule, type.Name));
+            }
+
+            /// <summary>
+            /// Checks rule: Ensure that {0} is declared as protected, virtual, and unsealed.
+            /// </summary>
+            private static void CheckDisposeBoolSignatureRule(IMethodSymbol method, INamedTypeSymbol type, SymbolAnalysisContext context)
+            {
+                if (method.DeclaredAccessibility != Accessibility.Protected ||
+                    !(method.IsVirtual || method.IsAbstract || method.IsOverride) || method.IsSealed)
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(DisposeBoolSignatureRule, $"{type.Name}.{method.Name}"));
+                }
+            }
+
+            /// <summary>
+            /// Checks rule: Modify {0} so that it calls Dispose(true), then calls GC.SuppressFinalize on the current object instance ('this' or 'Me' in Visual Basic), and then returns.
+            /// </summary>
+            private void CheckDisposeImplementationRule(IMethodSymbol method, INamedTypeSymbol type, ImmutableArray<IOperation> operationBlocks, OperationBlockAnalysisContext context)
+            {
+                var validator = new DisposeImplementationValidator(suppressFinalizeMethod, type);
+                if (!validator.Validate(operationBlocks))
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(DisposeImplementationRule, $"{type.Name}.{method.Name}"));
+                }
+            }
+
+            /// <summary>
+            /// Checks rule: Modify {0} so that it calls Dispose(false) and then returns.
+            /// </summary>
+            private static void CheckFinalizeImplementationRule(IMethodSymbol method, INamedTypeSymbol type, ImmutableArray<IOperation> operationBlocks, OperationBlockAnalysisContext context)
+            {
+                // TODO: Implement check of Finalize
```

Comment:
Yes currently there's a bug that operation blocks are not getting generated for finalizers in C#. It's tracked here - https://github.com/dotnet/roslyn/issues/7428.

Could you please add a testcase for this and skip it with this bug id - so that we know to implement this when the bug is fixed,